File: db4oj.tests/src/com/db4o/db4ounit/common/querying/QueryingVersionFieldTestCase.java
/* Copyright (C) 2008  Versant Inc.  http://www.db4o.com */

package com.db4o.db4ounit.common.querying;

import com.db4o.*;
import com.db4o.config.*;
import com.db4o.ext.*;
import com.db4o.query.*;

import db4ounit.*;
import db4ounit.extensions.*;


public class QueryingVersionFieldTestCase extends AbstractDb4oTestCase {
    
    public static void main(String[] arguments) {
        new QueryingVersionFieldTestCase().runAll();
    }
    
    public static class Item {
        
        public String name;
        
        public Item(String name_){
            name = name_;
        }
        
    }
    
    protected void configure(Configuration config) throws Exception {
        config.generateCommitTimestamps(true);
    }
    
    public void test(){
        
        storeItems(new String[] {"1", "2", "3"});
        db().commit();
        long initialTransactionVersionNumber = db().version();
        
        updateItem("2", "modified2");
        db().commit();
        long updatedTransactionVersionNumber = db().version();  
        
        Query q = db().query();
        q.constrain(Item.class);
        q.descend(VirtualField.COMMIT_TIMESTAMP).constrain(new Long(initialTransactionVersionNumber)).greater();
        
        // This part really isn't needed for this test case, but it shows, how changes
        // between two specific transaction commits can be queried.
        q.descend(VirtualField.COMMIT_TIMESTAMP).constrain(new Long(updatedTransactionVersionNumber)).smaller().equal();
        
        ObjectSet objectSet = q.execute();
        Assert.areEqual(1, objectSet.size());
        Item item = (Item) objectSet.next();
        Assert.areEqual("modified2", item.name);
        
        
        
    }

    private void updateItem(String originalName, String updatedName) {
        Item item = queryForItem(originalName);
        item.name = updatedName;
        store(item);
        
    }

    private Item queryForItem(String name) {
        Query q = newQuery(Item.class);
        q.descend("name").constrain(name);
        ObjectSet objectSet = q.execute();
        Assert.areEqual(1, objectSet.size());
        return (Item) objectSet.next();
    }

    private void storeItems(String[] names) {
        for (int i = 0; i < names.length; i++) {
            Item item = new Item(names[i]);
            store(item);
        }
    }

}

File: db4o.net/Db4objects.Db4o.Tests/Db4objects.Db4o.Tests/Common/Querying/QueryingVersionFieldTestCase.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using Db4oUnit;
using Db4oUnit.Extensions;
using Db4objects.Db4o;
using Db4objects.Db4o.Config;
using Db4objects.Db4o.Ext;
using Db4objects.Db4o.Query;
using Db4objects.Db4o.Tests.Common.Querying;

namespace Db4objects.Db4o.Tests.Common.Querying
{
	public class QueryingVersionFieldTestCase : AbstractDb4oTestCase
	{
		public static void Main(string[] arguments)
		{
			new QueryingVersionFieldTestCase().RunAll();
		}

		public class Item
		{
			public string name;

			public Item(string name_)
			{
				name = name_;
			}
		}

		/// <exception cref="System.Exception"></exception>
		protected override void Configure(IConfiguration config)
		{
			config.GenerateCommitTimestamps(true);
		}

		public virtual void Test()
		{
			StoreItems(new string[] { "1", "2", "3" });
			Db().Commit();
			long initialTransactionVersionNumber = Db().Version();
			UpdateItem("2", "modified2");
			Db().Commit();
			long updatedTransactionVersionNumber = Db().Version();
			IQuery q = Db().Query();
			q.Constrain(typeof(QueryingVersionFieldTestCase.Item));
			q.Descend(VirtualField.CommitTimestamp).Constrain(initialTransactionVersionNumber
				).Greater();
			// This part really isn't needed for this test case, but it shows, how changes
			// between two specific transaction commits can be queried.
			q.Descend(VirtualField.CommitTimestamp).Constrain(updatedTransactionVersionNumber
				).Smaller().Equal();
			IObjectSet objectSet = q.Execute();
			Assert.AreEqual(1, objectSet.Count);
			QueryingVersionFieldTestCase.Item item = (QueryingVersionFieldTestCase.Item)objectSet
				.Next();
			Assert.AreEqual("modified2", item.name);
		}

		private void UpdateItem(string originalName, string updatedName)
		{
			QueryingVersionFieldTestCase.Item item = QueryForItem(originalName);
			item.name = updatedName;
			Store(item);
		}

		private QueryingVersionFieldTestCase.Item QueryForItem(string name)
		{
			IQuery q = NewQuery(typeof(QueryingVersionFieldTestCase.Item));
			q.Descend("name").Constrain(name);
			IObjectSet objectSet = q.Execute();
			Assert.AreEqual(1, objectSet.Count);
			return (QueryingVersionFieldTestCase.Item)objectSet.Next();
		}

		private void StoreItems(string[] names)
		{
			for (int i = 0; i < names.Length; i++)
			{
				QueryingVersionFieldTestCase.Item item = new QueryingVersionFieldTestCase.Item(names
					[i]);
				Store(item);
			}
		}
	}
}
